### ISSUE TITLE
Support for OpenCV3

### ISSUE THREAD

**Issue by nikhilketkar** [beniz/deepdetect#36]
Linking CXX executable dede
/usr/bin/ld: ../src/libddetect.a(caffelib.cc.o): undefined reference to symbol '_ZN2cv6imreadERKNS_6StringEi'
/usr/local/lib/libopencv_imgcodecs.so.3.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
main/CMakeFiles/dede.dir/build.make:90: recipe for target 'main/dede' failed
make[2]: *** [main/dede] Error 1
CMakeFiles/Makefile2:160: recipe for target 'main/CMakeFiles/dede.dir/all' failed
make[1]: *** [main/CMakeFiles/dede.dir/all] Error 2
Makefile:76: recipe for target 'all' failed
make: *** [all] Error 2

**Comment by beniz** [comment]
Looks like a problem with OpenCV 3, can you use 2.x by any chance ? I've seen many reports of difficulties with OpenCV 3 and Caffe as well.

That said, if you are willing to test on something and see whether it clears your issue, can you try to add `opencv_imgcodecs` to `main/CMakeLists.txt` so the last line looks like:
```
target_link_libraries (dede ddetect glog gflags opencv_core opencv_highgui opencv_imgproc opencv_imgcodecs cppnetlib-uri curlpp curl crypto ssl ${Boost_LIBRARIES} ${CAFFE_LIB_DEPS})
```

This is similar to the changes reported in https://github.com/BVLC/caffe/pull/3186/files

Let me know how compilation goes after this change.

**Comment by az2012** [comment]
For me works! Add only target_link_libraries(fisherfaces ${the_library} opencv_core opencv_imgproc opencv_highgui opencv_imgcodecs)

**Comment by beniz** [comment]
Thanks, could you share the exact diff to the existing code?

**Comment by iamgroot42** [comment]
The problem is still persistent.

I am currently using Ubuntu 14.04, with openCV 3.0.0 and caffe 1.0.0r3

`.build_release/lib/libcaffe.so: undefined reference to `cv::imread(cv::String const&, int)'
.build_release/lib/libcaffe.so: undefined reference to `cv::imencode(cv::String const&, cv::_InputArray const&, std::vector<unsigned char, std::allocator<unsigned char> >&, std::vector<int, std::allocator<int> > const&)'
.build_release/lib/libcaffe.so: undefined reference to `cv::imdecode(cv::_InputArray const&, int)'
collect2: error: ld returned 1 exit status
make[3]: *** [.build_release/tools/caffe.bin] Error 1
make[3]: *** Waiting for unfinished jobs....
.build_release/lib/libcaffe.so: undefined reference to `cv::imread(cv::String const&, int)'
.build_release/lib/libcaffe.so: undefined reference to `cv::imencode(cv::String const&, cv::_InputArray const&, std::vector<unsigned char, std::allocator<unsigned char> >&, std::vector<int, std::allocator<int> > const&)'
.build_release/lib/libcaffe.so: undefined reference to `cv::imdecode(cv::_InputArray const&, int)'
collect2: error: ld returned 1 exit status
make[3]: *** [.build_release/tools/upgrade_net_proto_text.bin] Error 1
.build_release/lib/libcaffe.so: undefined reference to `cv::imread(cv::String const&, int)'
.build_release/lib/libcaffe.so: undefined reference to `cv::imencode(cv::String const&, cv::_InputArray const&, std::vector<unsigned char, std::allocator<unsigned char> >&, std::vector<int, std::allocator<int> > const&)'
.build_release/lib/libcaffe.so: undefined reference to `cv::imdecode(cv::_InputArray const&, int)'
collect2: error: ld returned 1 exit status
make[3]: *** [.build_release/examples/cifar10/convert_cifar_data.bin] Error 1
.build_release/lib/libcaffe.so: undefined reference to `cv::imread(cv::String const&, int)'
.build_release/lib/libcaffe.so: undefined reference to `cv::imencode(cv::String const&, cv::_InputArray const&, std::vector<unsigned char, std::allocator<unsigned char> >&, std::vector<int, std::allocator<int> > const&)'
.build_release/lib/libcaffe.so: undefined reference to `cv::imdecode(cv::_InputArray const&, int)'
collect2: error: ld returned 1 exit status
make[3]: *** [.build_release/tools/upgrade_net_proto_binary.bin] Error 1
.build_release/lib/libcaffe.so: undefined reference to `cv::imread(cv::String const&, int)'
.build_release/lib/libcaffe.so: undefined reference to `cv::imencode(cv::String const&, cv::_InputArray const&, std::vector<unsigned char, std::allocator<unsigned char> >&, std::vector<int, std::allocator<int> > const&)'
.build_release/lib/libcaffe.so: undefined reference to `cv::imdecode(cv::_InputArray const&, int)'
collect2: error: ld returned 1 exit status
make[3]: *** [.build_release/tools/compute_image_mean.bin] Error 1
.build_release/lib/libcaffe.so: undefined reference to `cv::imread(cv::String const&, int)'
.build_release/lib/libcaffe.so: undefined reference to `cv::imencode(cv::String const&, cv::_InputArray const&, std::vector<unsigned char, std::allocator<unsigned char> >&, std::vector<int, std::allocator<int> > const&)'
.build_release/lib/libcaffe.so: undefined reference to `cv::imdecode(cv::_InputArray const&, int)'
collect2: error: ld returned 1 exit status
make[3]: *** [.build_release/examples/mnist/convert_mnist_data.bin] Error 1
.build_release/lib/libcaffe.so: undefined reference to `cv::imread(cv::String const&, int)'
.build_release/lib/libcaffe.so: undefined reference to `cv::imencode(cv::String const&, cv::_InputArray const&, std::vector<unsigned char, std::allocator<unsigned char> >&, std::vector<int.build_release,/ libstd/:libcaffe.so::allocator <undefinedint >reference  >to  const`&cv)::imread(cv::String const&, int)'
.build_release/lib/'libcaffe.so
:. undefined reference to `cv::imencode(cv::String const&, cv::build_release_InputArray/ libconst&, std::vector<unsigned char, std::allocator<unsigned char> >&/,libcaffe.so :std ::undefinedvector <referenceint ,to  std::allocator<int> >` cvconst:&:)'imdecode
(.cvbuild_release:/:lib_InputArray/ libcaffe.soconst:& ,undefined  intreference) 'to
 `cv::imdecode(cv::_InputArray const&, int)'
collect2: error: ld returned 1 exit status
collect2: error: ld returned 1 exit status
make[3]: *** [.build_release/tools/upgrade_solver_proto_text.bin] Error 1
make[3]: *** [.build_release/tools/convert_imageset.bin] Error 1
.build_release/lib/libcaffe.so: undefined reference to `cv::imread(cv::String const&, int)'
.build_release/lib/libcaffe.so: undefined reference to `cv::imencode(cv::String const&, cv::_InputArray const&, std::vector<unsigned char, std::allocator<unsigned char> >&, std::vector<int, std::allocator<int> > const&)'
.build_release/lib/libcaffe.so: undefined reference to `cv::imdecode(cv::_InputArray const&, int)'
collect2: error: ld returned 1 exit status
make[3]: *** [.build_release/tools/extract_features.bin] Error 1
.build_release/examples/cpp_classification/classification.o: In function `main':
classification.cpp:(.text.startup+0x2d2): undefined reference to `cv::imread(cv::String const&, int)'
.build_release/lib/libcaffe.so: undefined reference to `cv::imencode(cv::String const&, cv::_InputArray const&, std::vector<unsigned char, std::allocator<unsigned char> >&, std::vector<int, std::allocator<int> > const&)'
.build_release/lib/libcaffe.so: undefined reference to `cv::imdecode(cv::_InputArray const&, int)'
collect2: error: ld returned 1 exit status
make[3]: *** [.build_release/examples/cpp_classification/classification.bin] Error 1
.build_release/lib/libcaffe.so: undefined reference to `cv::imread(cv::String const&, int)'
.build_release/lib/libcaffe.so: undefined reference to `cv::imencode(cv::String const&, cv::_InputArray const&, std::vector<unsigned char, std::allocator<unsigned char> >&, std::vector<int, std::allocator<int> > const&)'
.build_release/lib/libcaffe.so: undefined reference to `cv::imdecode(cv::_InputArray const&, int)'
collect2: error: ld returned 1 exit status
make[3]: *** [.build_release/examples/siamese/convert_mnist_siamese_data.bin] Error 1
make[2]: *** [caffe_dd/src/caffe_dd-stamp/caffe_dd-configure] Error 2
make[1]: *** [CMakeFiles/caffe_dd.dir/all] Error 2
make: *** [all] Error 2
`

**Comment by beniz** [comment]
We will escalate this one as OpenCV 3 needs to supported in a near future.

**Comment by beniz** [comment]
PR #125 has been tested, and should be merged soon, cc @iamgroot42 if you are willing to give it a try.

**Comment by beniz** [comment]
Merged #125
